### PR TITLE
Refactor `FocalPointPicker` to function component

### PR DESCRIPTION
**Pull request by stokesman**
Part of #22890.

When starting on this I first looked for some small things that could be simplified and put them in the first commit since they would be hard to spot within the rewrite as a function component.  Once I got into the rewrite, I found more things to simplify:
- Mainly that the logic and state handling the drag area boundaries was crufty from before #22531 where the image area was a fixed aspect and the drag area had variable margins around it (see https://github.com/WordPress/gutenberg/pull/22531#issuecomment-632824022).
- Additionally, there was some code for handling drag gestures that has been obviated by `useDragging` from the compose package.

## Testing Instructions
Make sure the `FocalPointPicker` works as expected. Especially in the Media & Text block and Cover blocks.

## Types of changes
Refactor

## Checklist:
- [x] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/components/src/focal-point-picker/controls.js [hunk 1]
```diff
@@ -22,19 +22,19 @@ const TEXTCONTROL_MAX = 100;
...
-	const valueY = fractionToPercentage( percentages.y );
+	const valueX = fractionToPercentage( point.x );
+	const valueY = fractionToPercentage( point.y );
 
 	const handleChange = ( value, axis ) => {
 		const num = parseInt( value, 10 );
 
 		if ( ! isNaN( num ) ) {
-			onChange( { ...percentages, [ axis ]: num / 100 } );
+			onChange( { ...point, [ axis ]: num / 100 } );
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/controls.js:33]: This is cosmetic but `point` just seems like a better name than `percentages`. It's an internal component so there's no concern of breakage.

File: packages/components/src/focal-point-picker/focal-point.js [hunk 2]
```diff
@@ -13,18 +13,12 @@ import {
...
+export default function FocalPoint( { left = '50%', top = '50%', ...props } ) {
 	const classes = classnames(
 		'components-focal-point-picker__icon_container'
 	);
 
-	const style = {
-		left: coordinates.left,
-		top: coordinates.top,
-	};
+	const style = { left, top };
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/focal-point.js:21]: Cosmetic. The separate `left` and `right` props are a touch simpler. Less abstract seems a fine thing for an internal component.

File: packages/components/src/focal-point-picker/grid.js [hunk 3]
```diff
@@ -16,11 +16,7 @@ import { useUpdateEffect } from '../utils/hooks';
 const { clearTimeout, setTimeout } =
 	typeof window !== 'undefined' ? window : {};
 
-export default function FocalPointPickerGrid( {
-	bounds = {},
-	value,
-	...props
-} ) {
+export default function FocalPointPickerGrid( { bounds, value, ...props } ) {
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/grid.js]: `bounds` is always supplied by the parent component and does not need to default.

File: packages/components/src/focal-point-picker/media.js [hunk 4]
```diff
@@ -18,7 +13,7 @@ export default function Media( {
 	alt,
 	autoPlay,
 	src,
-	onLoad = noop,
+	onLoad,
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/media.js:11]: There is no need to default this. The parent component supplies it unconditionally.

File: packages/components/src/focal-point-picker/media.js [hunk 5]
```diff
@@ -60,18 +55,14 @@ export default function Media( {
...
 	 * This async callback mimics the onLoad (img) / onLoadedData (video) callback
 	 * for media elements. It is used in the main <FocalPointPicker /> component
 	 * to calculate the dimensions + boundaries for positioning.
 	 */
 	useLayoutEffect( () => {
-		window.requestAnimationFrame( () => {
-			onLoadRef.current();
-		} );
+		window.requestAnimationFrame( () => onLoad() );
 	}, [] );
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/media.js]: There's no need to use a ref.

File: packages/components/src/focal-point-picker/test/index.js [hunk 6]
```diff
@@ -34,7 +34,8 @@ describe( 'FocalPointPicker', () => {
 		} );
 
 		it( 'should stop a drag operation when focus is lost', () => {
-			expect( firedDragEnd && ! firedDrag ).toBe( true );
+			expect( firedDrag ).toBe( undefined );
+			expect( firedDragEnd ).toBe( true );
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/test/index.js]: I wrote this test a while back. Seems better and more readable this way.

File: packages/components/src/focal-point-picker/utils.js [hunk 7]
```diff
@@ -1,10 +1,6 @@
 export const INITIAL_BOUNDS = {
-	top: 0,
-	left: 0,
-	bottom: 0,
-	right: 0,
-	width: 0,
-	height: 0,
+	width: 200,
+	height: 170,
 };
```
Comment (stokesman) [inline packages/components/src/focal-point-picker/utils.js:4]: The width and height values are extracted from the styles and the styles now pull them from here. The component was also no longer in need of all the other properties.

**Review by stokesman — COMMENTED** [review]
I've left some inline comments on some changes that are either stylistic or otherwise non-obvious.

**Comment by stokesman** [comment]
@mirka and/or @ciampo, I've had this sitting for a good while and I'd appreciate your advice on it. Specifically whether you'd prefer to see this also converted to TypeScript in this same PR (like I noticed was recently done with `FormTokenField`).

**Comment by mirka** [comment]
> Specifically whether you'd prefer to see this also converted to TypeScript in this same PR (like I noticed was recently done with `FormTokenField`).

What I took from my experience reviewing that PR was: It's probably easier to review if the FC and TS conversions are done in separate PRs 😄 I just didn't think it was worth codifying in the contributing guide because there are only a few class components left.

Is this ready for review, in that case?

File: packages/components/src/focal-point-picker/media.js (hunk 5 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/media.js]: Well there might have been a reason to use a ref so that exhaustive-deps wouldn't complain but this layout effect is removed in 7aedf1f in favor of an effect on the parent component.

**Comment by stokesman** [comment]
>Is this ready for review, in that case?

Now that I've linted with `react-hooks/exhaustive-deps` and added a couple commits to get it passing, I'll say yes.

File: packages/components/src/focal-point-picker/index.js [hunk 8]
```diff
@@ -129,7 +128,7 @@ export class FocalPointPicker extends Component {
 		}
 		return bounds;
 	}
-	updateValue( nextValue = {}, callback ) {
+	updateValue( nextValue, callback ) {
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js]: This change may introduce a regression for consumers of this component passing the `resolvePoint` callback, in case they expected that callback to never receive an `undefined` argument

File: packages/components/src/focal-point-picker/index.js [hunk 9]
```diff
@@ -182,7 +180,7 @@ export class FocalPointPicker extends Component {
 		const axis = keyCode === UP || keyCode === DOWN ? 'y' : 'x';
 		const value = parseFloat( next[ axis ] ) + delta;
 
-		next[ axis ] = roundClamp( value, 0, 1, step );
+		next[ axis ] = value;
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js]: What's the reason for this runtime change? (reviewing this change in the context of the first "cleanup" [commit](https://github.com/WordPress/gutenberg/pull/39168/commits/16271a8e808a70765a33791b8e49f5fb7b2e1612) in this PR)

File: packages/components/src/focal-point-picker/index.js [hunk 10]
```diff
@@ -223,10 +221,10 @@ export class FocalPointPicker extends Component {
...
 
 		// Enables holding shift to jump values by 10%
-		const step = byTenths ? 0.1 : 0.01;
-
-		nextX = roundClamp( nextX, 0, 1, step );
-		nextY = roundClamp( nextY, 0, 1, step );
+		if ( byTenths ) {
+			nextX = Math.round( nextX / 0.1 ) * 0.1;
+			nextY = Math.round( nextY / 0.1 ) * 0.1;
+		}
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js]: With this change, the values are not clamped anymore if `byTenths` is `false` — what's reason behind it? Why do are we moving away from `roundClamp` ? (reviewing this change in the context of the first "cleanup" [commit](https://github.com/WordPress/gutenberg/pull/39168/commits/16271a8e808a70765a33791b8e49f5fb7b2e1612) in this PR)

File: packages/components/src/focal-point-picker/index.js [hunk 11]
```diff
@@ -259,10 +257,7 @@ export class FocalPointPicker extends Component {
 		} = this.state;
 
 		if ( bounds.left === undefined || bounds.top === undefined ) {
-			return {
-				left: '50%',
-				top: '50%',
-			};
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js:270]: What's the reason behind this runtime change? (reviewing this change in the context of the first "cleanup" [commit](https://github.com/WordPress/gutenberg/pull/39168/commits/16271a8e808a70765a33791b8e49f5fb7b2e1612) in this PR)

File: packages/components/src/focal-point-picker/test/index.js [hunk 12]
```diff
@@ -76,12 +75,10 @@ describe( 'FocalPointPicker', () => {
...
 			);
-			// Click and press arrow up
 			const dragArea = getByRole( 'button' );
-			act( () => {
-				fireEvent.mouseDown( dragArea );
-				fireEvent.keyDown( dragArea, { charCode: 0, keyCode: 38 } );
-			} );
+			fireEvent.mouseDown( dragArea, { clientX: 1, clientY: 1 } );
+			// Change doesn't fire until mouseup
+			fireEvent.mouseUp( dragArea );
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/test/index.js]: What's the reason for this change in the unit tests?

File: packages/components/src/focal-point-picker/index.js [hunk 13]
```diff
@@ -73,13 +73,19 @@ export default function FocalPointPicker( {
...
+		const updateBounds = refUpdateBounds.current;
 		const { defaultView } = dragAreaRef.current.ownerDocument;
 		defaultView.addEventListener( 'resize', updateBounds );
-		return () => {
-			defaultView.removeEventListener( 'resize', updateBounds );
-		};
+		return () => defaultView.removeEventListener( 'resize', updateBounds );
 	}, [] );
 
+	useEffect( () => {
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js]: Should this hook be a `useLayoutEffect`, similarly to how this [used to be](https://github.com/WordPress/gutenberg/pull/39168/commits/9eb7fcc4e714a70794daafd35815b6988c506704#r827583298) in the `MediaPlaceholderElement` component

**Review by ciampo — COMMENTED** [review]
Thank you @stokesman for working on this.

I had a look at the component on Storybook and found a few regressions:

- the coordinates reset when the component loses focus (i.e. clicking outside of the interactive area)
- the thirds grid in the "Video" example does not adapt to its container

With regards to the TS refactor, I would have probably worked on that first, in order to gain even more confidence when working on the follow-up refactor from class to functional component.

I would also like to express a personal preference for avoiding runtime changes as much as possible (ref optimizations, drag logic refactor, variable name changes...) during "refactor" PRs like this one, as they increase considerably the difficulty amount of time required to review the PR. Those are all great improvements, but I would personally prefer if they were applied in one or more follow-up PRs, keeping the original PR focused on the refactor from class to function component.

File: packages/components/src/focal-point-picker/index.js (hunk 11 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js:270]: Just to remove redundancy because the only place `iconCoordinates` was called was to create the `coordinates` prop for the sub-component `FocalPoint` which defaults it to the same:
https://github.com/WordPress/gutenberg/blob/a9e8a8565e2c7bdf3e44a04462d96db5288428ff/packages/components/src/focal-point-picker/focal-point.js#L16-L17

File: packages/components/src/focal-point-picker/index.js (hunk 10 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: The values are clamped a few lines before this making the clamp of `roundClamp` redundant.

The larger story is one of yak-shaving. This PR got started when working on another PR (#34566) and wanting to modify `roundClamp` for a more specific purpose. I noticed the only other place it was used was this component, spotted the partial redundancy and figured it could be removed here to potentially free it up for the repurposing.

File: packages/components/src/focal-point-picker/index.js (hunk 9 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: This is a runtime change in the context of this commit but the next commit restores the clamping. The rounding isn't needed here because the increments are fixed and applied to values that are already rounded.

File: packages/components/src/focal-point-picker/index.js (hunk 8 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: I don't believe there’s a way this is called with `nextValue` as `undefined`. The closest equivalent to `updateValue` after the rewrite is `getFinalValue` (it calls `resolvePoint`) and I can double-check that to make sure there's no way it's being called with `undefined`.

File: packages/components/src/focal-point-picker/test/index.js (hunk 12 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/test/index.js]: Just that I recalled the test had originally been attempted with mouse events (https://github.com/WordPress/gutenberg/pull/38247#discussion_r803154354) and thought it'd good to try since it's probably the more expected way to test it. I think I'd also imagined it could help as a starting place in case a more a in-depth mouse based test was devised 🤷.

File: packages/components/src/focal-point-picker/index.js (hunk 13 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: I did do some more testing and found that could eliminate a visible reposition of the focal point. So it's done in 976b4b3. Maybe the isomorphic wrapper is extraneous because it's not easy for me to imagine this component being used in an SSR context.

**Comment by stokesman** [comment]
>-  the coordinates reset when the component loses focus (i.e. clicking outside of the interactive area)

🤦 That was introduced with 832ea2c and now fixed with 3d8d486

>- the thirds grid in the "Video" example does not adapt to its container

Great catch. Fixed it in 976b4b3. Sort of makes me think we may want a story with an intentionally 404'd media URL because otherwise we might have missed this. I don't believe a unit test could cover it because of jsdom and lack of element dimensions.

File: packages/components/src/focal-point-picker/index.js (hunk 13 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/index.js]: > Maybe the isomorphic wrapper is extraneous because it's not easy for me to imagine this component being used in an SSR context.

Maybe, but it feels more future-proof

File: packages/components/src/focal-point-picker/index.js [hunk 14]
```diff
@@ -23,317 +26,170 @@ import {
...
+
+	useEffect( () => {
+		const updateBounds = refUpdateBounds.current;
+		const { defaultView } = dragAreaRef.current.ownerDocument;
+		defaultView.addEventListener( 'resize', updateBounds );
+		return () => defaultView.removeEventListener( 'resize', updateBounds );
+	}, [] );
+
+	// Updates the bounds to cover cases of unspecified media or load failures.
+	useIsomorphicLayoutEffect( () => void refUpdateBounds.current(), [] );
```
Comment (ciampo) [inline packages/components/src/focal-point-picker/index.js:97]: Is there any reason for discarding the `if (!url)` check in the last commit?

File: packages/components/src/focal-point-picker/test/index.js (hunk 12 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/test/index.js]: Makes sense — thank you for the extra context!

File: packages/components/src/focal-point-picker/index.js (hunk 10 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/index.js]: Makes sense. I'm going to keep this conversation open for extra visibility in case anyone else reviews this PR

**Review by ciampo — COMMENTED** [review]
Most of my first round of feedback has been addressed, and the Storybook example seems to work as expected.

Given the fact that the code changes in this PR are not trivial, I've also tagged @mirka in case she wants to take a look as well before merging.

File: packages/components/src/focal-point-picker/index.js (hunk 14 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js:97]: When there is a `url` but it fails to load. It was what caused one of the regressions you noted:
>the thirds grid in the "Video" example does not adapt to its container

**Comment by ciampo** [comment]
While giving this PR another round of testing, I discovered another bug (which is also present in `trunk`):

Basically, `0` coordinates are displayed as `50` in the `UnitControl` inputs (for both left and top coordinates)

https://user-images.githubusercontent.com/1083581/179955982-b2eaa4f5-8644-438b-9588-ef4b491d50ff.mp4

**Comment by stokesman** [comment]
> I discovered another bug (which is also present in `trunk`)

Nice spotting that! Turns out it has the same cause as #42455. I tested by applying the changes from #42484 and it fixes it.

File: packages/components/src/focal-point-picker/index.js [hunk 15]
```diff
@@ -23,317 +26,170 @@ import {
...
-		 */
-		const {
-			isDragging,
-			percentages: { x, y },
-		} = this.state;
-		const { value } = this.props;
-		if ( ! isDragging && ( value.x !== x || value.y !== y ) ) {
-			this.setState( { percentages: this.props.value } );
-		}
-	}
```
Comment (mirka) [inline packages/components/src/focal-point-picker/index.js:87]: Are these no longer necessary? I didn't find equivalent logic in the refactored code.

**Comment by ciampo** [comment]
> > I discovered another bug (which is also present in `trunk`)
> 
> Nice spotting that! Turns out it has the same cause as #42455. I tested by applying the changes from #42484 and it fixes it.

Great timing! Let's prioritise #42484 — we can then rebase this PR after #42484 merged for a final round of testing

File: packages/components/src/focal-point-picker/index.js [hunk 16]
```diff
@@ -23,317 +26,170 @@ import {
...
+					onKeyDown={ arrowKeyStep }
+					onMouseDown={ startDrag }
+					onBlur={ () => {
+						if ( isDragging ) endDrag();
+					} }
+					ref={ dragAreaRef }
+					role="button"
+					tabIndex="-1"
+				>
+					<Grid bounds={ bounds } value={ `${ x }${ y }` } />
```
Comment (mirka) [inline packages/components/src/focal-point-picker/index.js]: (Just a comment, not meant to address in this PR:) I found it a bit strange that a prop named `value` is being used basically as a re-rendering key. I wonder if it would be more straightforward (and efficient) if we simply passed `isDragging` instead of every changed `value`.

File: packages/components/src/focal-point-picker/index.js [hunk 17]
```diff
@@ -23,317 +26,170 @@ import {
...
+		onDragMove: ( event ) => {
+			// Prevents text-selection when dragging.
+			event.preventDefault();
+			const value = getValueWithinDragArea( event );
+			onDrag?.( value, event );
+			keepPoint( value );
+		},
+		onDragEnd: ( event ) => {
+			onDragEnd?.( event );
+			sendPoint();
```
Comment (mirka) [inline packages/components/src/focal-point-picker/index.js]: I'm trying to understand the reason we need to maintain `point` and `refPoint` separately. Is it just for the `sendPoint()` on this line? In that case, couldn't we do a `const value = getValueWithinDragArea( event )` here as well?

**Review by mirka — COMMENTED** [review]
I don't think I've gone as deep as Marco did, but I read through the code and nothing major is jumping out 👍 

One happy find is that I think this PR fixes a bug in `trunk` where the indicator offset gets weird after scrolling:

https://user-images.githubusercontent.com/555336/180282036-e89b4559-e88a-4a74-ad19-2ec5399de673.mp4

File: packages/components/src/focal-point-picker/index.js (hunk 15 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js:87]: It's good to cover this. I'm glad you asked. The near-equivalent logic is here:
https://github.com/WordPress/gutenberg/blob/e0015da798a8fb7db50c1ebd51eb2db2efabee9f/packages/components/src/focal-point-picker/index.js#L69-L70
The internal state is used only while dragging and otherwise the incoming `value` prop is used (i.e. it’s controlled so things like UNDO/REDO will work).

I did decide the logic to stop dragging when the `url` prop changes was unnecessary based on testing and finding it to work fine (I can only imagine this happening due to UNDO/REDO and seems a quite obscure edge case).

I had tested all this months ago in the post editor and now to be sure I put together a story and retested.

<details><summary>Story for controllable <code>FocalPointPicker</code> using <kbd>Meta</kbd>+<kbd>x</kbd> to make controlled swaps of media and <kbd>Meta</kbd>+<kbd>z</kbd> to make controlled updates to the value</summary>

```diff
diff --git a/packages/components/src/focal-point-picker/stories/index.js b/packages/components/src/focal-point-picker/stories/index.js
index 475029611a..eec67b82a2 100644
--- a/packages/components/src/focal-point-picker/stories/index.js
+++ b/packages/components/src/focal-point-picker/stories/index.js
@@ -1,7 +1,7 @@
 /**
  * WordPress dependencies
  */
-import { useState } from '@wordpress/element';
+import { useEffect, useRef, useState } from '@wordpress/element';
 /**
  * Internal dependencies
  */
@@ -31,6 +31,49 @@ export const _default = () => {
 	return <Example />;
 };
 
+const Controllable = ( { x = 0.5, y = 0.5 } ) => {
+	const [ focalPoint, setFocalPoint ] = useState( { x, y } );
+	const [ media, setMedia ] = useState( 0 );
+	const refMediaList = useRef( [
+		`data:image/svg+xml,%3Csvg width='280' height='210' xmlns='http://www.w3.org/2000/svg'/%3E`,
+		`data:image/svg+xml,%3Csvg width='480' height='360' xmlns='http://www.w3.org/2000/svg'/%3E`,
+	] );
+	const refHeldKeySet = useRef( new Set() );
+	useEffect( () => {
+		const heldKeySet = refHeldKeySet.current;
+		const downer = ( { key } ) => {
+			heldKeySet.add( key );
+			if ( heldKeySet.has( 'Meta' ) ) {
+				if ( key === 'x' )
+					setMedia(
+						( v ) => ( v + 1 ) % refMediaList.current.length
+					);
+				else if ( key === 'z' )
+					setFocalPoint( ( v ) => ( {
+						x: 1 + v.x * -1,
+						y: 1 + v.y * -1,
+					} ) );
+			}
+		};
+		const upper = ( { key } ) => refHeldKeySet.current.delete( key );
+		document.addEventListener( 'keydown', downer );
+		document.addEventListener( 'keyup', upper );
+		return () => {
+			document.removeEventListener( 'keydown', downer );
+			document.removeEventListener( 'keyup', upper );
+		};
+	}, [] );
+	return (
+		<FocalPointPicker
+			value={ focalPoint }
+			onChange={ setFocalPoint }
+			url={ refMediaList.current[ media ] }
+		/>
+	);
+};
+
+export const controllable = () => <Controllable />;
+
 export const image = () => {
 	const url =
 		'https://i0.wp.com/themes.svn.wordpress.org/twentytwenty/1.3/screenshot.png?w=572&strip=al';

```
</details>

Here is a demo of controlled updates switching the media while dragging and afterward changing the value.

https://user-images.githubusercontent.com/9000376/181604876-9dab9da2-0c22-4cc5-8a88-8cead98677a8.mp4

File: packages/components/src/focal-point-picker/index.js (hunk 16 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: Yes, 💯 strange.

Your suggested solution seems great. It will require a little reworking of `Grid` but I'd expect it to be an all around improvement.

File: packages/components/src/focal-point-picker/index.js (hunk 17 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: I'd wanted to optimize by not recalculating through `getValueWithinDragArea` because there is no change of pointer coordinates with this event. On the other hand, I also don't like having both ref and state (I'd previously had another way to avoid that but it also had its ugliness).

I'm open to changing this as you've posited. I can't imagine there would be a perceivable performance difference.

**Comment by stokesman** [comment]
>One happy find is that I think this PR fixes a bug in trunk where the indicator offset gets weird after scrolling:

I vaguely remember encountering that and I definitely remember simplifying the logic to find the proper offset. Yet I also think that issue is only in Storybook for some reason. Maybe the iframe? So perhaps it could happen in the Site Editor as well (EDIT: probably not because I think only the canvas is in the iframe and not the inspector controls). I can't remember how much I looked into that.

**Comment by ciampo** [comment]
Hey @stokesman , do you think you'll be able to finish the work on this one?

File: packages/components/src/focal-point-picker/index.js (hunk 16 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: I thought of something that might block the suggestion. Right now `Grid` makes itself visible not just when dragging but also during any value changes (arrow keys while the drag area is focused or changes at the number inputs). There'd be no way to maintain that by merely using `isDragging`.

Maybe just changing the prop from `value` to `key` would be the way to go.

File: packages/components/src/focal-point-picker/index.js (hunk 17 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: In e0015da, I've removed `refPoint` and found it there was no need to recalculate the point `onDragEnd` either. Seems like in an earlier iteration I'd ran into some issue with the non-synchronous nature of the `point` and `isDragging` states but can't reproduce any of that now. Anyway, it's sweet to simplify this further!

**Comment by stokesman** [comment]
> do you think you'll be able to finish the work on this one?

Eventually 😄 Though, where I'd left off there were no requested changes and it seemed like a changelog was all that was required. Since then a couple merges introduced conflicts and I've rebased and resolved those. I also added a commit regarding the conversation Lena started https://github.com/WordPress/gutenberg/pull/39168#discussion_r926955477.

File: packages/components/src/focal-point-picker/index.js (hunk 16 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/index.js]: We could simplify the code by passing a `showOverlay` prop to the `Grid` component, moving the timeout logic to `index.js`

<details>
<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/focal-point-picker/grid.js b/packages/components/src/focal-point-picker/grid.js
index ecd6d6b7d9..ed6d8ae51d 100644
--- a/packages/components/src/focal-point-picker/grid.js
+++ b/packages/components/src/focal-point-picker/grid.js
@@ -1,8 +1,3 @@
-/**
- * WordPress dependencies
- */
-import { useState } from '@wordpress/element';
-
 /**
  * Internal dependencies
  */
@@ -11,21 +6,16 @@ import {
 	GridLineX,
 	GridLineY,
 } from './styles/focal-point-picker-style';
-import { useUpdateEffect } from '../utils/hooks';
-
-export default function FocalPointPickerGrid( { bounds, value, ...props } ) {
-	const animationProps = useRevealAnimation( value );
-	const style = {
-		width: bounds.width,
-		height: bounds.height,
-	};
 
+export default function FocalPointPickerGrid( { bounds, ...props } ) {
 	return (
 		<GridView
 			{ ...props }
-			{ ...animationProps }
 			className="components-focal-point-picker__grid"
-			style={ style }
+			style={ {
+				width: bounds.width,
+				height: bounds.height,
+			} }
 		>
 			<GridLineX style={ { top: '33%' } } />
 			<GridLineX style={ { top: '66%' } } />
@@ -34,25 +24,3 @@ export default function FocalPointPickerGrid( { bounds, value, ...props } ) {
 		</GridView>
 	);
 }
-
-/**
- * Custom hook that renders the "flash" animation whenever the value changes.
- *
- * @param {string} value Value of (box) side.
- */
-function useRevealAnimation( value ) {
-	const [ isActive, setIsActive ] = useState( false );
-
-	useUpdateEffect( () => {
-		setIsActive( true );
-		const timeout = window.setTimeout( () => {
-			setIsActive( false );
-		}, 600 );
-
-		return () => window.clearTimeout( timeout );
-	}, [ value ] );
-
-	return {
-		isActive,
-	};
-}
diff --git a/packages/components/src/focal-point-picker/index.js b/packages/components/src/focal-point-picker/index.js
index 09128c04b9..5c1f594641 100644
--- a/packages/components/src/focal-point-picker/index.js
+++ b/packages/components/src/focal-point-picker/index.js
@@ -27,6 +27,9 @@ import {
 	MediaContainer,
 } from './styles/focal-point-picker-style';
 import { INITIAL_BOUNDS } from './utils';
+import { useUpdateEffect } from '../utils/hooks';
+
+const GRID_OVERLAY_TIMEOUT = 600;
 
 export default function FocalPointPicker( {
 	autoPlay = true,
@@ -45,6 +48,7 @@ export default function FocalPointPicker( {
 	},
 } ) {
 	const [ point, setPoint ] = useState( valueProp );
+	const [ showGridOverlay, setShowGridOverlay ] = useState( false );
 
 	const { startDrag, endDrag, isDragging } = useDragging( {
 		onDragStart: ( event ) => {
@@ -146,6 +150,15 @@ export default function FocalPointPicker( {
 	const instanceId = useInstanceId( FocalPointPicker );
 	const id = `inspector-focal-point-picker-control-${ instanceId }`;
 
+	useUpdateEffect( () => {
+		setShowGridOverlay( true );
+		const timeout = window.setTimeout( () => {
+			setShowGridOverlay( false );
+		}, GRID_OVERLAY_TIMEOUT );
+
+		return () => window.clearTimeout( timeout );
+	}, [ x, y ] );
+
 	return (
 		<BaseControl
 			label={ label }
@@ -165,7 +178,7 @@ export default function FocalPointPicker( {
 					role="button"
 					tabIndex="-1"
 				>
-					<Grid bounds={ bounds } value={ `${ x }${ y }` } />
+					<Grid bounds={ bounds } showOverlay={ showGridOverlay } />
 					<Media
 						alt={ __( 'Media preview' ) }
 						autoPlay={ autoPlay }
diff --git a/packages/components/src/focal-point-picker/styles/focal-point-picker-style.js b/packages/components/src/focal-point-picker/styles/focal-point-picker-style.js
index 0687275c7f..58fa3eb5de 100644
--- a/packages/components/src/focal-point-picker/styles/focal-point-picker-style.js
+++ b/packages/components/src/focal-point-picker/styles/focal-point-picker-style.js
@@ -61,7 +61,6 @@ export const ControlWrapper = styled( Flex )`
 
 export const GridView = styled.div`
 	left: 50%;
-	opacity: 0;
 	overflow: hidden;
 	pointer-events: none;
 	position: absolute;
@@ -70,11 +69,7 @@ export const GridView = styled.div`
 	transition: opacity 120ms linear;
 	z-index: 1;
 
-	${ ( { isActive } ) =>
-		isActive &&
-		`
-		opacity: 1;
-	` }
+	opacity: ${ ( { showOverlay } ) => ( showOverlay ? 1 : 0 ) };
 `;
 
 export const GridLine = styled.div`
```
</details>

File: packages/components/src/focal-point-picker/index.js (hunk 15 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/index.js:87]: Tested the story, works as expected 👌

File: packages/components/src/focal-point-picker/index.js (hunk 17 above)
Comment (ciampo) [reply packages/components/src/focal-point-picker/index.js]: Makes sense to me, the last known value from the `dragmove` event should be already good.

Thank you for the explanation!

**Comment by ciampo** [comment]
Thank you, Mitchell! I gave another round of test in Storybook and the component is looking and behaving well!

Once https://github.com/WordPress/gutenberg/pull/39168#discussion_r952441301 is addressed and a CHANGELOG entry is added, we should be able to merge it!

File: packages/components/src/focal-point-picker/index.js (hunk 16 above)
Comment (stokesman) [reply packages/components/src/focal-point-picker/index.js]: Works for me! I applied it in 4f99530. I also tested again in Storybook and the Post Editor and found it works as expected.

**Review by ciampo — APPROVED** [review]
Great job! 🚀

**Comment by stokesman** [comment]
Thanks for testing and reviewing Marco and Lena 🙇
